drivers/sensor/lis2dh/lis2dh.c

```diff
@@ -165,13 +166,215 @@ static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
 		odr--;
 	}
 
-	return data->hw_tf->write_reg(dev, LIS2DH_REG_CTRL1,
+	/* store last set freq value so it can be restored */
+	if (freq != lis2dh->target_odr) {
+		lis2dh->target_odr = odr;
+		LOG_DBG("set target odr to %i", lis2dh->target_odr);
+	}
+
+	/* If sensor is powered down only set the target frequency 
+	 * but don't apply it to hardware, re-enabling the sensor
+	 * with lis2dh_power_down() will do that.
+	 */
+#if defined(CONFIG_PM_DEVICE)
+	if (lis2dh->pm_state == PM_DEVICE_STATE_LOW_POWER) {
+		LOG_INF("New ODR value will be applied once the sensor enters active mode.");
+		return 0;
+	}
+#endif
+	return lis2dh->hw_tf->write_reg(dev, LIS2DH_REG_CTRL1,
 				      (value & ~LIS2DH_ODR_MASK) |
 				      LIS2DH_ODR_RATE(odr));
 }
 #endif
 
-#ifdef CONFIG_LIS2DH_ACCEL_RANGE_RUNTIME
+#if (defined(CONFIG_LIS2DH_OPER_MODE_RUNTIME) || defined(CONFIG_LIS2DH_OPER_MODE_HIGH_RES))
+static int lis2dh_set_high_res(const struct device *dev, bool enable)
+{
+	struct lis2dh_data *lis2dh = dev->data;
+	int ret;
+	uint8_t value;
+
+	if (enable) {
+		/* re-enable High resolution mode */
+		ret = lis2dh->hw_tf->read_reg(dev, LIS2DH_REG_CTRL4, &value);
+		if (unlikely(ret < 0)) {
+			return ret;
+		}
+		return lis2dh->hw_tf->write_reg(dev, LIS2DH_REG_CTRL4, value | LIS2DH_HR_BIT);
+	} else { /* disable */
+		ret = lis2dh->hw_tf->read_reg(dev, LIS2DH_REG_CTRL4, &value);
+		if (unlikely(ret < 0)) {
+			return ret;
+		}
+		return lis2dh->hw_tf->write_reg(dev, LIS2DH_REG_CTRL4, value & ~LIS2DH_POWER_DOWN);
+	}
+}
+#endif /* CONFIG_LIS2DH_OPER_MODE_RUNTIME || CONFIG_LIS2DH_OPER_MODE_HIGH_RES */
+
+#if defined(CONFIG_LIS2DH_OPER_MODE_RUNTIME)
+int lis2dh_set_oper_mode(const struct device *dev, enum lis2dh_oper_mode new_mode)
+{
+	struct lis2dh_data *lis2dh = dev->data;
+	int ret;
+
+	if (new_mode == lis2dh->oper_mode){
+		return 0;
+	}
+
+	LOG_INF("New oper mode: %i", (uint8_t)new_mode);
+
+#if defined(CONFIG_PM_DEVICE)
+	if (lis2dh->pm_state == PM_DEVICE_STATE_LOW_POWER) {
+		lis2dh->oper_mode = new_mode;
+		LOG_INF("New operation mode will be applied once the sensor enters active mode.");
+		return 0;
+	}
+#endif
+
+	if (new_mode == OPER_MODE_LOW_POWER){
+		if (lis2dh->oper_mode == OPER_MODE_HIGH_RES) {
+			ret = lis2dh_set_high_res(dev, false);
+			if (unlikely(ret < 0)) {
+				return ret;
+			}
+		}
+		ret = lis2dh->hw_tf->update_reg(dev, LIS2DH_REG_CTRL1,
+									(uint8_t)LIS2DH_LP_EN_BIT_MASK,
```

Comment:
I would take care on your formatting a bit. This extra tabs are a bit too much

Reply:
My IDE displays tabs shorter then github. Ill remove some tabs.